### ISSUE TITLE
cluster-health reports healthy with contradictory output

### ISSUE THREAD

**Issue by stresler** [coreos/etcd#7032]
I don't see how this can be a logical output: 

```
/ # etcdctl --ca-file /etc/kubernetes/ssl/ca.pem  --endpoint https://etcd0.DOMAIN.com:2379 cluster-health
member 7641137365309009 is unhealthy: got unhealthy result from https://etcd1.DOMAIN.com:2379
member 7e51c10fd2e88c9c is unreachable: no available published client urls
member abeec58727dc4111 is healthy: got healthy result from https://etcd2.DOMAIN.com:2379
member c5decd33b4e6bcf9 is unhealthy: got unhealthy result from https://etcd0.DOMAIN.com:2379
member fb7b0b7455c4e5df is unreachable: no available published client urls
cluster is healthy
```

**Comment by xiang90** [comment]
If one of the node reports healthy, then your cluster is healthy. 

It is possible that the your machine cannot reach some of the nodes. It does not mean that these node are unhealthy themselves. This is actually what happened in your case. There are two unreachable nodes from etcdctl. You should probably fix that.

**Comment by stresler** [comment]
No, I'm just now setting it up, and these aren't unexpected results at the moment. 

But, really? A cluster is healthy if only one node is healthy? That seems like either the cluster-health report does something besides report cluster health, or it needs a new name. 

If 4/5 nodes aren't healthy how can it honestly say the 'cluster' is healthy?

**Comment by stresler** [comment]
To be clearer, I would expect cluster-health to do the first part - which is seems to do, then give an assessment - e.g. "1 of 5 nodes is healthy, cluster is sub-optimal" or some such.

**Comment by xiang90** [comment]
> 1 of 5 nodes is healthy, cluster is sub-optimal

The cluster itself might be totally fine. But a user runs etcdctl on a machine that can only access one of the nodes due to firewall or other client sides issues. What should we report? 

Right now, if one node reports healthy, which means the cluster can make progress from its POV, the cluster is healthy. This is current if we define the cluster is health if the cluster can accept and process request.

If you want to define cluster healthy as ALL members are healthy, then we might see false positive as the case I mentioned above.

We can be smarter for sure, but already given the detailed output, I am not sure if we want. If you want to improve the output, please submit a pr or the approach you want.

**Comment by heyitsanthony** [comment]
@xiang90 are you sure? It looks like `member c5decd33b4e6bcf9 is unhealthy: got unhealthy result from https://etcd0.DOMAIN.com:2379` only happens if `etcdctl` gets a response from the member.

**Comment by stresler** [comment]
I mean... I don't really want to argue the point. 

If you're reporting 'cluster healthy' when you can't actually know that - due to firewalls or whatever - then you aren't reporting accurate output and that is a bug. 

You could report 'cluster state unknowable.' and be accurate, I suppose. But to make an assertion that the cluster is healthy directly following output that potentially contradicts that is a bug.

**Comment by xiang90** [comment]
@heyitsanthony @stresler 

The cluster healthy thing is HARD to get 100% right, and lacks of thought. That is why we removed it in etcdctl v3. 

In v2 etcdctl, we define the cluster is healthy if the cluster can make progress during the checking. If you want to change the definition, what should we want to change to? The cluster is health only if all can make a progress? This seems wrong.

**Comment by xiang90** [comment]
@stresler 

> If you're reporting 'cluster healthy' when you can't actually know that - due to firewalls or whatever - then you aren't reporting accurate output and that is a bug.

I do not think you understand what I said.

We currently define that if the cluster made any progress during the health checking, the cluster is healthy.

Then if etcdctl observes the cluster made progress from any etcd member it checks, then the cluster is healthy based on the definition.

If you suggest us to change the definition, can you tell me what is your exact expectation? And how you would like to define the definition?

**Comment by stresler** [comment]
I'm new to this project, and mostly just as a user. Can you tell me what you mean by 'make progress'? This isn't a term I'm familiar with.  

As a new user, I would define 'healthy' as 'in sync or on track to be in sync with the leader and accessible as advertised'... presumably if not, that would require some sort of corrective action to make the node healthy. 

I'm sure there is nuance, or potentially other states - but 'making progress' doesn't tell me much. Making progress towards what?

**Comment by stresler** [comment]
In this case the message for 2/5 is clear - node unreachable.  That means, as the cluster understands it needs the remaining 3/5 to be healthy to have consensus... it can't verify that, so I would think the result would be 'cluster not healthy' or 'health unverifiable'.

**Comment by xiang90** [comment]
@stresler 

Made progress means the cluster accepted a user request, got it committed through the majority of the cluster, and replied a response to a client. Basically the cluster finished one full request/response.

**Comment by xiang90** [comment]
> In this case the message for 2/5 is clear - node unreachable. That means, as the cluster understands it needs the remaining 3/5 to be healthy to have consensus...

That is the client got confused, not the cluster itself. The client cannot connect does not mean the cluster cannot connect to these member internally.

I do not think it is correct to say "cluster is unhealthy" if a client could not connect to some nodes when the cluster can still make progress.

I think the definition you expect is 

Only report cluster is healthy if etcdctl can verify cluster health from every nodes. 

Then some users will argue they want etcdctl to report the other way: if the majority of the cluster is alive, the cluster should be thought as healthy. I do not know.

So we already remove the cluster healthy output completely from v3 output.

**Comment by heyitsanthony** [comment]
@xiang90 disregarding however it's defined in v2 and only judging by the output, does it look "healthy"? Two members have unreachable client addresses, two members can't make progress, and only one thinks it's fine. For a point of comparison, suppose you have a RAID6 and two drives are toast, do you want your RAID to happily chirp "healthy"?

It could stand to be stricter, at least by taking unhealthy responses seriously. Right now it's either the cluster makes progress and is therefore healthy or it is completely down and therefore unhealthy. Going by this health analogy, isn't it more useful to know you're sick before waiting until you drop dead?

> Then some users will argue they want etcdctl to report the other way: if the majority of the cluster is alive, the cluster should be thought as healthy.

Sure, but I'm going to claim those users are wrong to call it that. I agree that definition makes sense for per-member/endpoint health checking, but when considering the cluster as a whole (and if we add a cluster-health to ctlv3), that attitude seems somewhat confused. If any member responds it's severed from raft, the cluster may be "online" or "available", but it's hardly "healthy" without the fault tolerance from that member.

**Comment by xiang90** [comment]
> For a point of comparison, suppose you have a RAID6 and two drives are toast, do you want your RAID to happily chirp "healthy"?

I do not say it is the best way to define "healthy". It is the expected output based on the current definition.

> Going by this health analogy, isn't it more useful to know you're sick before waiting until you drop dead?

I agree the whole thing (cluster healthy) lacks clear definition.

Neither

- 1 member healthy -> cluster healthy 

nor

- only member healthy from the client POV -> cluster healthy

make sense to me.

I feel that simply giving the output from each member checking to users is good enough. Thus we have endpoint status instead of cluster health command in v3.

I do not really want etcdctl to interpret the meaning of the combined output to generate one simple sentence, which will be hard to understand without providing additional context.

If you want to have that cluster healthy output and ensure it make sense, I would suggest that we document its meaning somewhere before move forward.

**Comment by heyitsanthony** [comment]
@xiang90 OK, can this be reopened? The current definition (or lack thereof, the definition appears to be the way etcdctl behaves, the docs seem to only say bad client urls won't affect the health) is clearly inadequate. Making it stricter so it respects unhealthy responses won't break anything that's not already broken, as far as I can tell.

**Comment by xiang90** [comment]
@heyitsanthony 

We can:

1. ignore the unreachable.
2. if all members reports health -> healthy
3. no members report healthy -> unhealthy
4. some reports healthy, others not -> cluster still works, but need to be fixed.

**Comment by vhosakot** [comment]
I have etcd up on three nodes.  When I kill etcd on two nodes, I see `etcdctl` reports `cluster is unhealthy`.

etcd is up on `mcp1.paslab009007.mc.metacloud.in` and killed on `mcp2.paslab009007.mc.metacloud.in` and `mcp3.paslab009007.mc.metacloud.in`.

Below, I see `etcdctl` shows `got unhealthy result from` for the node on which etcd is up (`mcp1.paslab009007.mc.metacloud.in`), and `is unreachable` for the two nodes on which etcd is killed (`mcp2.paslab009007.mc.metacloud.in` and `mcp3.paslab009007.mc.metacloud.in`).

```
# etcdctl --endpoints=http://mcp1.paslab009007.mc.metacloud.in:2379,http://mcp2.paslab009007.mc.metacloud.in:2379,http://mcp3.paslab009007.mc.metacloud.in:2379 cluster-health
member 5a51fea559eb7b83 is unhealthy: got unhealthy result from http://mcp1.paslab009007.mc.metacloud.in:2379
failed to check the health of member b90ed230bf96a611 on http://mcp3.paslab009007.mc.metacloud.in:2379: Get http://mcp3.paslab009007.mc.metacloud.in:2379/health: dial tcp 172.18.7.162:2379: getsockopt: connection refused
member b90ed230bf96a611 is unreachable: [http://mcp3.paslab009007.mc.metacloud.in:2379] are all unreachable
failed to check the health of member f8b67df9eb695a18 on http://mcp2.paslab009007.mc.metacloud.in:2379: Get http://mcp2.paslab009007.mc.metacloud.in:2379/health: dial tcp 172.18.7.161:2379: getsockopt: connection refused
member f8b67df9eb695a18 is unreachable: [http://mcp2.paslab009007.mc.metacloud.in:2379] are all unreachable
cluster is unhealthy

# etcdctl --endpoints=http://mcp1.paslab009007.mc.metacloud.in:2379,http://mcp2.paslab009007.mc.metacloud.in:2379,http://mcp3.paslab009007.mc.metacloud.in:2379 member list
Failed to get leader:  client: etcd cluster is unavailable or misconfigured

# curl http://mcp1.paslab009007.mc.metacloud.in:2379/health
{"health": "false"}

# curl http://mcp2.paslab009007.mc.metacloud.in:2379/health
curl: (7) Failed connect to mcp2.paslab009007.mc.metacloud.in:2379; Connection refused

# curl http://mcp3.paslab009007.mc.metacloud.in:2379/health
curl: (7) Failed connect to mcp3.paslab009007.mc.metacloud.in:2379; Connection refused
```

Is this expected behavior that etcd will not work on just a single node in a multi-node cluster?  Is it because etcd's leader election fails with just a single node up as it is a single point of failure in the cluster?

In a multi-node cluster, is there way to make etcd work even on a single node on a best-effort basis even if it is a single point of failure and risky?

**Comment by stresler** [comment]
This is actually expected. You need enough nodes to form consensus. In a 3 node cluster that would be 2 nodes minimum.

**Comment by vhosakot** [comment]
cool, thanks for the reply @stresler! 😄

**Comment by heyitsanthony** [comment]
fixed by #8070. closing
